### PR TITLE
cw1-subkeys: Implement permission functionality

### PR DESCRIPTION
**Pull request by orkunkl**
Implements permission functionality for messages:

- Delegate
- Undelegate
- Redelegate
- Withdraw

### CODE REVIEW

File: packages/cw0/src/expiration.rs
Comment (orkunkl) [inline packages/cw0/src/expiration.rs:18]: Useful for debugging

File: contracts/cw1-subkeys/src/contract.rs
Comment (orkunkl) [inline contracts/cw1-subkeys/src/contract.rs:111]: My only question is this @ethanfrey

File: contracts/cw1-subkeys/README.md
Comment (ethanfrey) [inline contracts/cw1-subkeys/README.md:41]: Maybe `StakingPermissions` as a better name? Or `SetStakingPermissions`?

File: contracts/cw1-subkeys/README.md
Comment (ethanfrey) [inline contracts/cw1-subkeys/README.md:52]: You removed the `AllowanceResponse`? I think you need both.

Oh, this is README....

File: contracts/cw1-subkeys/helpers.ts
Comment (ethanfrey) [inline contracts/cw1-subkeys/helpers.ts:138]: Question:

If we add permissions for eg. governance votes, or calling wasm contracts, would you keep them here? Or use a separate type? 

To see if we want one large "Permissions" or "StakingPermissions", "VotingPermissions", etc.

File: contracts/cw1-subkeys/src/state.rs
Comment (ethanfrey) [inline contracts/cw1-subkeys/src/state.rs:20]: Nice

File: packages/cw0/src/expiration.rs
Comment (ethanfrey) [reply packages/cw0/src/expiration.rs:18]: Makes better error messages, right? sounds good.

File: contracts/cw1-subkeys/helpers.ts
Comment (ethanfrey) [inline contracts/cw1-subkeys/helpers.ts:281]: Thanks for adding this here as well.

I think the `all_allowances` query is missing as well (my bad), can you add that too?

File: contracts/cw1-subkeys/helpers.ts
Comment (ethanfrey) [inline contracts/cw1-subkeys/helpers.ts:162]: I would rather make this tighter check;

`type CosmosMsg = SendMsg | DelegateMsg | UndelegateMsg | RedelegateMsg | WithdrawMsg`

File: contracts/cw1-subkeys/src/contract.rs
Comment (ethanfrey) [inline contracts/cw1-subkeys/src/contract.rs:93]: This makes a quite large branch.
I would do something like:

`CosmosMsg::Staking(staking_msg) => check_staking_message(&staking_msg, &mut allowance)?` 

or such and then include all these branches inside `check_staking_message`. That function would then be simpler:

```rust
match staking_msg {
  StakingMsg:::Delegate { amount, ..} => 
                    if !allowance.permissions.delegate {
                        Err(StdError::generic_err(PermissionErr::Delegate{}))
                        // or maybe this (and cast to StdError above when `?` is used)
                        Err(PermissionErr::Delegate{})
                    } else {
                       Ok(())
                   }
}
```

Try to pull some out and simplify

File: contracts/cw1-subkeys/src/contract.rs
Comment (ethanfrey) [inline contracts/cw1-subkeys/src/contract.rs:101]: I would keep the staking items independent of balances. You can delegate and undelegate all you want, but not take money from the account. By requiring the balances here, I need to give someone the power to take my money to let them select the validators I invest in. Just the first check is needed (and see above comment about refactor to make it simpler)

File: contracts/cw1-subkeys/src/contract.rs
Comment (ethanfrey) [inline contracts/cw1-subkeys/src/contract.rs:260]: We should be able to set the staking permissions on accounts without any balances. 

It may just have balances, or just permissions, or both (and they should not interfere with each other)

File: contracts/cw1-subkeys/src/contract.rs
Comment (ethanfrey) [inline contracts/cw1-subkeys/src/contract.rs:417]: I would not add permissions here. The setup script is already too confusing for me. And this makes it harder to test individual cases. (Just set balance, just set permissions, set both). I guess they are somehow covered here, but it should be more explicit.

I was already a bit unhappy with this giant setup script and adding more makes it super hard to see what is being tested. I would rather just use this for a few of the targetted tests below, and just call `init` and a few Allowance/Permssions settings explicitly for the tests you are writing

File: contracts/cw1-subkeys/src/contract.rs
Comment (ethanfrey) [inline contracts/cw1-subkeys/src/contract.rs:1102]: Good to have some non-defaults here... let's pick a name or two.

File: contracts/cw1-subkeys/src/contract.rs
Comment (ethanfrey) [inline contracts/cw1-subkeys/src/contract.rs:1098]: `HumanAddr::from("validator")` does the same and is less typing

**Review by ethanfrey — COMMENTED** [review]
Thanks for your first contract PR @orkunkl 

In general, you got all the pieces connected and it looks generally correct. I added a number of comments to improve code quality. Can you go through them and also take another pass on naming design and I will then re-review this

File: contracts/cw1-subkeys/README.md
Comment (orkunkl) [reply contracts/cw1-subkeys/README.md:41]: `SetStakingPermission` sounds good

File: contracts/cw1-subkeys/src/contract.rs
Comment (orkunkl) [reply contracts/cw1-subkeys/src/contract.rs:260]: On it

File: contracts/cw1-subkeys/helpers.ts
Comment (orkunkl) [reply contracts/cw1-subkeys/helpers.ts:138]: Separate permissions design brings a lot of extra code(bucket) to both smart contract and helpers.ts. This struct will not contain hundreds of booleans(max 15-20?). I guess using single permissions struct is the optimal way

Comment (orkunkl) [reply contracts/cw1-subkeys/helpers.ts:138]: If rust and buckets provides a way to define permissions generically, then multiple permissions that extend `Permissions` trait is better.

File: contracts/cw1-subkeys/README.md
Comment (orkunkl) [reply contracts/cw1-subkeys/README.md:41]: Depends on https://github.com/CosmWasm/cosmwasm-plus/pull/75/files#r484310769

File: contracts/cw1-subkeys/README.md
Comment (orkunkl) [reply contracts/cw1-subkeys/README.md:52]: I wanted to make helpers.ts and msg.rs naming identical

File: packages/cw0/src/expiration.rs
Comment (orkunkl) [reply packages/cw0/src/expiration.rs:18]: Exactly

File: contracts/cw1-subkeys/helpers.ts
Comment (ethanfrey) [reply contracts/cw1-subkeys/helpers.ts:142]: Fair enough, we keep one Permissions and extend it for other packages.
Please add a comment on the type explaining that choice (actually only in helpers.rs, not here)

**Comment by ethanfrey** [comment]
Looks good. One comment requested, and please `cargo fmt` so the linter is happy. Then I will do the final review to merge

File: contracts/cw1-subkeys/helpers.ts
Comment (orkunkl) [reply contracts/cw1-subkeys/helpers.ts:142]: I assume you mean state.rs by helpers.rs? or helpers.ts? I am confused 😕

Comment (ethanfrey) [reply contracts/cw1-subkeys/helpers.ts:142]: I meant `state.rs`
I should re-review again

File: contracts/cw1-subkeys/helpers.ts
Comment (ethanfrey) [inline contracts/cw1-subkeys/helpers.ts:177]: :+1:

File: contracts/cw1-subkeys/src/contract.rs
Comment (ethanfrey) [inline contracts/cw1-subkeys/src/contract.rs:90]: This is fine and explicit. I generally make this more compact.

```rust
let perm = permissions(&mut deps.storage).load(owner_raw.as_slice())?;
check_staking_permissions(staking_msg, perm)?;
```

But it is fine as is. (and with custom error message)

File: contracts/cw1-subkeys/src/contract.rs
Comment (ethanfrey) [inline contracts/cw1-subkeys/src/contract.rs:129]: Much nicer

File: contracts/cw1-subkeys/src/contract.rs
Comment (ethanfrey) [inline contracts/cw1-subkeys/src/contract.rs:138]: You can ignore the attributes like: 

```rust
StakingMsg::Undelegate { .. } => { 
            if !permissions.undelegate {
                return Err(PermissionErr::Undelegate {});
            }
}
```

File: contracts/cw1-subkeys/src/contract.rs
Comment (ethanfrey) [inline contracts/cw1-subkeys/src/contract.rs:277]: `s/Cannot set allowance/Cannot set permission/`

File: contracts/cw1-subkeys/src/contract.rs
Comment (ethanfrey) [inline contracts/cw1-subkeys/src/contract.rs:306]: Nice addition

File: contracts/cw1-subkeys/src/contract.rs
Comment (ethanfrey) [inline contracts/cw1-subkeys/src/contract.rs:368]: Interesting chain. 

I think the `?` and `match` approach for allowances is more legible. But nice that you learned a bit more about map and map_or and handling Options and Results. Key skills for working with rust

File: contracts/cw1-subkeys/src/contract.rs
Comment (ethanfrey) [inline contracts/cw1-subkeys/src/contract.rs:696]: Looks good. I would add one more test, where you set the allowance for someone with a permission and then ensure that both are set properly (setting allowance doesn't unset the permission)

File: contracts/cw1-subkeys/src/contract.rs
Comment (ethanfrey) [inline contracts/cw1-subkeys/src/contract.rs:1319]: I would add a few more tests where one sender has eg. 2 permissions set, 2 unset and ensure that those two pass, and the other 2 fail. 

Just a double check that there is no typo in which permission is checked for which message. And ensures future changes don't break this (what if we swap delegate and redelegate permission checks? these tests would still pass)

**Review by ethanfrey — COMMENTED** [review]
Much nicer.

If you could cover a few more cases in the test code, I can merge this

File: contracts/cw1-subkeys/src/contract.rs
Comment (orkunkl) [reply contracts/cw1-subkeys/src/contract.rs:368]: Just some old scala and functional programming skills λ

Comment (orkunkl) [reply contracts/cw1-subkeys/src/contract.rs:368]: Applied the `?` and `match` case.
But `can_send` method looks weird to me. The return value is `StdResult` but in tests, it is expected to be always boolean. Which throws the error information to the thrash 🤔

File: contracts/cw1-subkeys/src/contract.rs
Comment (orkunkl) [reply contracts/cw1-subkeys/src/contract.rs:1319]: Added the test. Developer errors must be more distinctive and functional than just a str :( https://github.com/CosmWasm/cosmwasm/issues/518

File: contracts/cw1-subkeys/src/contract.rs
Comment (ethanfrey) [reply contracts/cw1-subkeys/src/contract.rs]: Error should only throw if we have trouble parsing from disk, and we cannot *prove* that will never happen.

In all normal cases (non-corrupt db), we should get true or false. You would have to write invalid json to the key to trigger an error

File: contracts/cw1-subkeys/src/contract.rs
Comment (ethanfrey) [inline contracts/cw1-subkeys/src/contract.rs:359]: This could actually be:

`Some(permission) => Ok(check_staking_permissions(&staking_msg, permission).is_ok())`

Error only returned from `may_load` unable to parse.

File: contracts/cw1-subkeys/src/contract.rs
Comment (ethanfrey) [inline contracts/cw1-subkeys/src/contract.rs:1246]: This shadows the function `coin` for the rest of this test function. I would avoid that 

(I only shadow one variable with another when transforming it, eg. `let msg = msg.to_string();`)

File: contracts/cw1-subkeys/src/contract.rs
Comment (ethanfrey) [inline contracts/cw1-subkeys/src/contract.rs:1340]: You want typed errors... I know we will make this happen

File: contracts/cw1-subkeys/src/contract.rs
Comment (ethanfrey) [inline contracts/cw1-subkeys/src/contract.rs:1370]: Thank you.

**Review by ethanfrey — APPROVED** [review]
Looks good.

Two minor nits on code style. You can ideally change them (or ignore them if you disagree strongly) and then ready for merge.

File: contracts/cw1-subkeys/src/contract.rs
Comment (orkunkl) [reply contracts/cw1-subkeys/src/contract.rs:1246]: I see :+1:  Renamed it to `coin1`. `allow` was not a good fit there.

**Comment by orkunkl** [comment]
2 points left:
1. `StdErr` which is taken care of. We can keep this PR on hold until it is there. It's a minor improve so good to merge IMO.
2. This is important. In helpers.ts, [wasm source URL](https://github.com/CosmWasm/cosmwasm-plus/blob/e88048b51d5b7a7ccfc8b448847943f85a1cc75a/contracts/cw1-subkeys/helpers.ts#L352) is pointed to v0.1.1 tag. After merge, cli helper will be broken. It is a good practice to use tags in `helpers.ts` files but to make it work we need to release new version after every contract change.... I don't see a problem with releasing minor versions. @ethanfrey ideas?

**Comment by ethanfrey** [comment]
`helpers.ts` is a bit of an issue. current flow is this:

1. Create 0.2.1 tag of rust code
2. Package is auto-built (20 minutes later)
3. Update helpers.ts to point to new code binary

Meaning 0.2.1 helpers.ts points to 0.2.0 (or 0.1.1) code.

I am happy to make a new tag shortly and then we have the updated helpers, but this chicken-and-egg problem is what stops us from using tagged helpers.ts, which is the real solution. Any ideas for work-arounds?

**Comment by ethanfrey** [comment]
I am happy to merge this when you give the :+1:

And we can cut a release of cosmwasm-plus contract today

**Comment by orkunkl** [comment]
Houston, the shuttle is ready to launch 🚀  👍

**Pull request merged**
